### PR TITLE
Replace BOOST_STATIC_ASSERT with static_assert

### PR DESCRIPTION
**Pull request by str4d**
Closes zcash/zcash#4823.

### CODE REVIEW

**Comment by str4d** [comment]
To verify this, check out the branch and then run:
```
./contrib/devtools/commit-script-check.sh master..HEAD
```

File: contrib/devtools/commit-script-check.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#!/bin/sh
+# Copyright (c) 2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
Comment (daira) [inline contrib/devtools/commit-script-check.sh:4]:
```suggestion
# file COPYING or https://www.opensource.org/licenses/mit-license.php .
```

**Review by daira — APPROVED** [review]
utACK

**Comment by rex4539** [comment]
```
/Users/rex/rex4539-zcash-boost on branch 4823-static-assert rex@MacBook-Pro-2018% ./contrib/devtools/commit-script-check.sh master..HEAD
sed: 1: "/^-BEGIN VERIFY SCRIPT- ...": unexpected EOF (pending }'s)
Error: missing script for: 3b3f4f0484afe5ac31f7ce14a0f558df5e501ea9
Failed
```

**Comment by str4d** [comment]
@rex4539 could you test this again? I just force-pushed with some extra changes that include portability fixes.

**Comment by str4d** [comment]
At @daira's request I've moved the checker into #4827. Once that is merged I'll rebase this PR so it only contains the Boost changes.

**Comment by rex4539** [comment]
Compiles fine on native macOS and all tests pass.

**Comment by str4d** [comment]
Rebased on master now that #4827 has merged.

**Review by daira — APPROVED** [review]
Checked utACK.

**Comment by daira** [comment]
@zkbot r+

**Pull request merged**